channeldb/migration27/migration.go

```diff
@@ -0,0 +1,117 @@
+	"fmt"
+
+	mig "github.com/lightningnetwork/lnd/channeldb/migration_01_to_11"
+	"github.com/lightningnetwork/lnd/kvdb"
+)
+
+var (
+	// historicalChannelBucket stores all channels that have seen their
+	// commitment tx confirm. All information from their previous open state
+	// is retained.
+	historicalChannelBucket = []byte("historical-chan-bucket")
+)
+
+// MigrateHistoricalBalances patches the two new fields, `InitialLocalBalance`
+// and `InitialRemoteBalance`, for all the open channels saved in historical
+// channel bucket. Unlike migration 25, it will only read the old channel info
+// first and then patch the new tlv records with empty values. For historical
+// channels, we previously didn't save the initial balances anywhere and since
+// it's corresponding open channel bucket is deleted after closure, we have
+// lost that balance info.
+func MigrateHistoricalBalances(tx kvdb.RwTx) error {
+	log.Infof("Migrating historical local and remote balances...")
+
+	// First fetch the top level bucket which stores all data related to
+	// historically stored channels.
+	rootBucket := tx.ReadWriteBucket(historicalChannelBucket)
+
+	// If no bucket is found, we can exit early.
+	if rootBucket == nil {
+		return nil
+	}
+
+	// Read a list of historical channels.
+	channels, err := findHistoricalChannels(rootBucket)
+	if err != nil {
+		return err
+	}
+
+	// Migrate the balances.
+	for _, c := range channels {
+		if err := migrateBalances(rootBucket, c); err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+// findHistoricalChannels finds all historical channels.
+func findHistoricalChannels(historicalBucket kvdb.RBucket) ([]*OpenChannel,
+	error) {
+
+	channels := []*OpenChannel{}
+
+	// readChannel is a helper closure that reads the channel info from the
+	// historical sub-bucket.
+	readChannel := func(rootBucket kvdb.RBucket, cp []byte) error {
+		c := &OpenChannel{}
+
+		// Read the sub-bucket.
+		chanBucket := rootBucket.NestedReadBucket(cp)
+		if chanBucket == nil {
+			log.Errorf("unable to read bucket for chanPoint=%x", cp)
+			return nil
+		}
+
+		// Get the old channel info.
+		if err := FetchChanInfo(chanBucket, c, true); err != nil {
+			return fmt.Errorf("unable to fetch chan info: %v", err)
+		}
+
+		channels = append(channels, c)
+
+		return nil
+	}
+
+	// Iterate the root bucket.
+	err := historicalBucket.ForEach(func(cp, _ []byte) error {
+		return readChannel(historicalBucket, cp)
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return channels, nil
+}
+
+// migrateBalances serializes the channel info using the new tlv format where
+// the two fields, `InitialLocalBalance` and `InitialRemoteBalance` are patched
+// with empty values.
+func migrateBalances(rootBucket kvdb.RwBucket, c *OpenChannel) error {
+	var chanPointBuf bytes.Buffer
+	err := mig.WriteOutpoint(&chanPointBuf, &c.FundingOutpoint)
+	if err != nil {
+		return err
+	}
+
+	// Get the channel bucket.
+	chanBucket, err := rootBucket.CreateBucketIfNotExists(
```

Comment:
Why `CreateBucketIfNotExists` vs just grabbing the bucket and exiting if it doesn't exist? AFAICT at this point we already know that this channel point is _definitely_ in the bucket as well?

Reply:
Yeah exactly, fixed!